R-inline-resiliency-policies-in-pubsub-subscribers.md

```diff
@@ -0,0 +1,181 @@
+# Inline resiliency policies in pubsub subscriber declarations
+
+* Author(s): Oliver Tomlinson (@olitomlinson)
+* State: Draft
+* Updated: 18/11/2023
+
+## Overview
+
+I propose that declarative and programmatic PubSub subscribers should be extended specify a resiliency policy which ONLY affects the subscriber in question.
```

Comment:
Define `subscriber in question` please. This is very ambiguous.

The pubsub component + app ID is a single subscriber in question.

Essentially this would override any resiliency policy for the component with regards to that very specific sidecar instance!

If you also want to include something topic specific here, this becomes a huge undertaking (requires refactoring of components, completely changing the policies we have etc)

Loading policies that are component specific from an app to overload the resiliency configuration may be possible however!

Reply:
> Define `subscriber in question` please. This is very ambiguous. 

Hmmm... Maybe 'subscription' rather than 'subscriber' would be better.

"I propose that declarative and programmatic PubSub subscriptions should be extended to include a resiliency policy which is applied only to the the inbound process of delivering the message from sidecar to app. The resiliency policy is scoped to just the subscription it was declared into / against"

Does that read better?

Reply:
I'm proposing a change which doesn't have to impact any of the existing component resiliency mechanics. It simply swaps out the existing component policy (if one exists) and applies a more specialised policy (if one has been provided via `dapr/subscribe` or via a subscription yaml) at the time of creating the subscription onto the topic.

<img width="1286" alt="image" src="https://github.com/dapr/proposals/assets/4224880/56f44bca-18df-4ce8-8c34-ddd661eff0ab">

Reply:
Yes but still a bit too ambiguous.

1) Would these settings apply to all topics within this subscription?

Or 2) do you need to be able to configure different settings for each topic?

The latter would be an extraordinary amount of work, and could instead be accomplished by repeating 1) with a different app and configuration - but still using the same pubsub component.

Reply:
I didn't see this image earlier. As long as **subscription** specific resiliency configuration applies to all topics within that subscription this sounds good to me!

Reply:
1. No. Each subscription has the ability to provide its own unique policy, _just_ for that subscription. They are not shared across each Topic.

The whole point of this proposal is about enabling each subscription to operate via its own specialised policy, if the component-level policy is not suitable.

> I didn't see this image earlier.

Ive just created the diagram to aid our conversation :)

> As long as subscription specific resiliency configuration applies to all topics within that subscription this sounds good to me!

No, unfortunately it doesn't apply to all Topics as a blanket policy, it only applies to the subscriptions that that have been configured with the inline policy.

Reply:
I'll ping you in Discord because I think you are still misunderstanding what I'm asking :)

Reply:
I'm talking about this difference:

```python
@app.route('/dapr/subscribe', methods=['GET'])
def subscribe():
    subscriptions = [
      {
        'pubsubname': 'pubsub',
        'topic': 'checkout',
        'routes': {
          'rules': [
            {
              'match': 'event.type == "order"',
              'path': '/orders'
            },
          ],
          'default': '/orders',
          'resiliency': {
           // Resiliency Config here
          }
        }
      },
      {
        'pubsubname': 'pubsub',
        'topic': 'checkout2',
        'routes': {
          'rules': [
            {
              'match': 'event.type == "order"',
              'path': '/orders2'
            },
          ],
          'default': '/orders2',
          'resiliency': {
           // SAME Resiliency Config here
          }
        }
      }
]
    return jsonify(subscriptions)
```

instead of

```python
@app.route('/dapr/subscribe', methods=['GET'])
def subscribe():
    subscriptions = [
      {
        'pubsubname': 'pubsub',
        'topic': 'checkout',
        'routes': {
          'rules': [
            {
              'match': 'event.type == "order"',
              'path': '/orders'
            },
          ],
          'default': '/orders',
          'resiliency': {
           // Resiliency Config here
          }
        }
      },
      {
        'pubsubname': 'pubsub',
        'topic': 'checkout2',
        'routes': {
          'rules': [
            {
              'match': 'event.type == "order"',
              'path': '/orders2'
            },
          ],
          'default': '/orders2',
          'resiliency': {
           // Different resiliency config here
          }
        }
      }
]
    return jsonify(subscriptions)
```


We can only accept one configuration per Pubsub component / Sidecar combination.